### PR TITLE
[FIX] mass_mailing, web: error handling of invalid leaf

### PR DESCRIPTION
**Pull request by sauc-odoo**
When a user enters an invalid domain it raises a ValueError of Invalid leaf. The error occurs everywhere where `widget="domain"` is used.

Steps to reproduce:
    1. Turn on developer mode.
    2. Install mass_mailing module and open it.
    3. Click on the new button.
    4. Enter subject and set recipients as contact or mailing contact.
    5. In code editor pass the domain as [('', '=', 0)]
    6. error will occur - ValueError: Invalid leaf ('', '=', 0)

See this traceback:
```
ValueError: Invalid leaf ('', '=', 1)
  File "odoo/http.py", line 2115, in __call__
    response = request._serve_db()
  File "odoo/http.py", line 1698, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "odoo/service/model.py", line 134, in retrying
    result = func()
  File "odoo/http.py", line 1725, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "odoo/http.py", line 1922, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "addons/website/models/ir_http.py", line 234, in _dispatch
    response = super()._dispatch(endpoint)
  File "odoo/addons/base/models/ir_http.py", line 154, in _dispatch
    result = endpoint(**request.params)
  File "odoo/http.py", line 715, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "addons/web/controllers/dataset.py", line 28, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "addons/web/controllers/dataset.py", line 24, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "odoo/api.py", line 457, in call_kw
    result = _call_kw_model(method, model, args, kwargs)
  File "odoo/api.py", line 430, in _call_kw_model
    result = method(recs, *args, **kwargs)
  File "odoo/models.py", line 1485, in search_count
    query = self._search(domain, limit=limit)
  File "odoo/addons/base/models/res_partner.py", line 899, in _search
    return super()._search(domain, offset, limit, order, access_rights_uid)
  File "odoo/models.py", line 4759, in _search
    query = self._where_calc(domain)
  File "odoo/models.py", line 4506, in _where_calc
    return expression.expression(domain, self).query
  File "odoo/osv/expression.py", line 448, in __init__
    self.parse()
  File "odoo/osv/expression.py", line 613, in parse
    push(leaf, self.root_model, self.root_alias)
  File "odoo/osv/expression.py", line 600, in push
    check_leaf(leaf, internal)
  File "odoo/osv/expression.py", line 397, in check_leaf
    raise ValueError("Invalid leaf %s" % str(element))
```

Similarly, this issue can be produced in different modules where `widget="domain"` is present.

After applying this commit the user will not be allowed to save an invalid domain.

sentry-3945982195

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Julien00859 — COMMENTED** [review]
Seems a better idea to me, I wonder if we can achieve the same without that extra field but with api.constraints instead

**Comment by sauc-odoo** [comment]
Hello @Julien00859 

We have used a computed field to display the number of records filtered out when the domain is applied(check here - https://drive.google.com/file/d/13tKzGSZlLcJaGmHQMB0HFgjDl20AXe8k/view?usp=sharing) and can raise ValidationError for bad domains simultaneously. Previously the number of records was displayed using RPC call which is removed in current PR.

We can apply api.constrains but we will not able to get number of records.

So kindly give your suggestions on what to do.

**Comment by Julien00859** [comment]
@sauc-odoo I don't get it why you would need to change how retrieving the record count behave...

Saving a record with a bad domain and using that domain to retrieve how many records there are are two separates operations. When the domain is bad, you shouldn't be able to save the record, thus the question about retrieving the records count is not relevant since the bad domain shouldn't had been saved.

I don't get it why you would do the two things at the same time. Plus saving the result of `search_count` on the record is a bad idea since it won't account for new records created on the related model that would match the domain.

I don't know how the js-framework works very well but here it seems that you added a mandatory field on all domain widgets but you only reflected that change on mass-mailing in the backend, not on the other models that use that domain widget too. The failing runbot tests seem to confirm my feelings.

So I do continue to think that it should be an api.constraint

**Review by flvr-odoo — CHANGES_REQUESTED** [review]

File: addons/mass_mailing/models/mailing.py [hunk 1]
```diff
@@ -234,6 +235,15 @@ def _get_default_mail_server_id(self):
         'The A/B Testing Percentage needs to be between 0 and 100%'
     )]
 
+    @api.constrains('mailing_model_real', 'mailing_domain')
+    def _check_mailing_domain(self):
+        try:
+            for mailing in self:
+                domain = ast.literal_eval(mailing.mailing_domain)
+                self.env[mailing.mailing_model_real].search_count(domain)
```
Comment (Julien00859) [inline addons/mass_mailing/models/mailing.py:243]: No need to actually do the query on postgres, simply crafting the `Query` object should be enough to sanity check the domain
```suggestion
                self.env[mailing.mailing_model_real]._search(domain)
```

File: addons/web/static/src/views/fields/domain/domain_field.js [hunk 2]
```diff
@@ -114,7 +114,7 @@ export class DomainField extends Component {
         let recordCount;
         const context = this.getContext(props);
         try {
-            recordCount = await this.orm.silent.searchCount(resModel, domain, { context });
+            recordCount = props.record.data[props.countField];
```
Comment (Julien00859) [inline addons/web/static/src/views/fields/domain/domain_field.js:117]: seems to be dead code

**Comment by sauc-odoo** [comment]
Hello @Julien00859 

Changes have been done, kindly review them.
Thanks.

File: addons/web/static/src/views/fields/domain/domain_field.js [hunk 3]
```diff
@@ -112,17 +112,6 @@ export class DomainField extends Component {
         }
 
         let recordCount;
-        const context = this.getContext(props);
```
Comment (Polymorphe57) [inline addons/web/static/src/views/fields/domain/domain_field.js:115]: Hello. You cannot remove this. recordCount is left undefined. This is a change of spec that is not wanted. This seems also totally unrelated to your original issue. Could you explain why add a constraint with _check_mailing_domain is not enough?

**Comment by sauc-odoo** [comment]
Hello @Polymorphe57 

Firstly, recordCount code was there in my solution and I created a `search_count_silent(self, domain)` method in `models.py` which was used to silent the error caused by `search_count` method. (check here - https://github.com/odoo/odoo/commit/c7b9392c42368272db6c5dff0eb2161f8f263b70).

But later on, I was asked to add a constraint that should stop users from saving bad domains. So to achieve it I had to remove recordCount code because that code was calling `search_count` method before constraint see below https://github.com/odoo/odoo/blob/a0f6bcb046ca66737099fa7f211b345e32241cf5/addons/web/static/src/views/fields/domain/domain_field.js#L99-L104  and causing the traceback in terminal. So if we want to keep the constraint we have to remove recordCount code and the constraint should also be added in each and every module where `widget="domain"` is used.

So kindly give your suggestions on what to do? :smile:

**Comment by Polymorphe57** [comment]
I have just tried with the constraint only and the problem does not appear anymore. Again I don't think you need more than that. Anyway, the search count feature is too important to be dropped.

**Comment by sauc-odoo** [comment]
The issue has been fixed in following PR:
https://github.com/odoo/odoo/pull/112326
https://github.com/odoo/odoo/pull/127699

**Pull request closed without merging**